torch/lib/THC/THCTensorScatterGather.cu

```diff
@@ -92,7 +92,7 @@ __global__ void THCudaTensor_gatherKernel(
                                                           tensor, &tensorOffset,
                                                           src, &srcOffset);
 
-    IndexType indexValue = (IndexType)index.data[indexOffset] - TH_INDEX_BASE;
+    long indexValue = index.data[indexOffset] - TH_INDEX_BASE;
```

Comment:
Hmm how about this:

```
assert(index.data[indexOffset] >= TH_INDEX_BASE)
IndexType indexValue = index.data[indexOffset] - TH_INDEX_BASE;
assert(index.data < tensor.sizes[dim]);
...
```

Because the index is guaranteed to fit in IndexType, and we can assert non-negative before storing? I'm not 100% sure how this would work in terms of the code it would generate.